File: python/boj1654.py
import sys
def binary_search(lan,n,start,end): #이분 탐색
    result = 0 #최댓값을 저장할 변수
    while start<=end: 
        mid = (start+end)//2 #평균(가운데의 값)
        cnt = 0 #랜선의 갯수 세기
        for i in lan:
            cnt += i // mid #랜선의 갯수를 평균으로 나눈 값 합하기
        if cnt >= n: #필요한 랜선의 갯수보다 자른 랜선의 갯수가 많거나 같으면
            start = mid + 1 #탐색 범위 변경
            result = mid #지금까지의 최댓값 저장
        else: #아니라면
            end = mid - 1 # 탐색 범위 변경
    return result #가장 마지막 최댓값 반환
    
k, n = map(int, sys.stdin.readline().split(" "))
lan = [int(sys.stdin.readline()) for _ in range(k)]
start = 1
end = max(lan) #가장 긴 랜선의 길이가 end
print(binary_search(lan,n,start,end))

File: C++/boj1654.cpp
#include <iostream>

int main(){
    int k, n;
    std::cin >> k >> n;
    std::cin.ignore();
    unsigned int* lan = new unsigned int[10000];
    int max = 0;

    for (int i = 0; i < k; i++){
        int num;
        std::cin >> num;
        lan[i] = num;
        if (max < num) max = num; // 이 if 문을 비트 연산으로 해보고 싶었는데 아쉽다
    }

    unsigned int start = 1;
    unsigned int end = max;
    unsigned int ans = 0;

    while (start <= end){
        unsigned int mid = (start + end) >> 1; // 비트 시프트로 나누기 2
        long long cnt = 0;
        for (int i = 0; i < k; i++){
            cnt += lan[i] / mid; // 동적 배열은 for each문 못쓰나? 일단 이렇게 해봄
        }
        if (cnt >= n){
            ans = mid;
            start = mid + 1;
        }
        else end = mid - 1;
    }
    std::cout << ans << std::endl;
    delete lan;
    // 시간복잡도는 O(n log n)? - 이분탐색 + 반복문
}